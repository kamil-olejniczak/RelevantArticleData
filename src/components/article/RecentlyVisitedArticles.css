.RecentlyVisitedArticles {
    margin: 0 auto;
    padding: 0 10px;
}

.RecentlyVisitedArticles tr {
    text-align: center;
}

.header {
    display: inline-block;
    font-weight: 700;
    margin-bottom: 10px;
    letter-spacing: 2px;
}

#resetStats {
    float: right;
    text-decoration: underline;
    color: blue;
    cursor: pointer;
}