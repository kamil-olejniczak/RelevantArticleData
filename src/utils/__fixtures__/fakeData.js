export const firstRecentlyVisitedSite = {
    id: 1,
    title: 'SomeTitle',
    url: 'SomeUrl',
    date: new Date(1 << 30),
};
export const secondRecentlyVisitedSite = {
    id: 2,
    title: 'SomeTitle2',
    url: 'SomeUrl2',
    date: new Date(1 << 30),
};
export const fakeRecentlyVisited = {
    recentlyVisited: [firstRecentlyVisitedSite, secondRecentlyVisitedSite]
};
export const fakeArticleData = {
    data: {
        title: 'Some Title',
        date_published: '1970-01-01',
        domain: 'www.dabliu.com',
        dek: 'Short dek text',
        excerpt: 'Short excerpt text',
        word_count: 111,
        lead_image_url: 'http://www.dabliu.pl/image.html',
    }
};
export const fakeTagStatistics = {
    div: 8,
    p: 2,
    a: 3,
    span: 10,
    i: 9,
    html: 1,
    br: 2,
    table: 7,
    hr: 6,
    input: 4,
    textarea: 5
};
export const fakeSortedTagStatistics = [
    'span', '10',
    'i', '9',
    'div', '8',
    'table', '7',
    'hr', '6',
    'textarea', '5',
    'input', '4',
    'a', '3',
    'br', '2',
    'p', '2',
];