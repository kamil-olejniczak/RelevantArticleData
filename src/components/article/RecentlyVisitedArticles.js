import React from 'react';
import {Link} from 'react-router-dom';
import './RecentlyVisitedArticles.css';

const RecentlyVisitedArticles = (props) => (
    <div className="RecentlyVisitedArticles">
        <span className="header">Recently Visited Articles:</span>
        <span id="resetStats" onClick={props.clearData}>Reset stats</span>
        {props.recentlyVisited.map(article => (
            <div key={article.id}>
                <Link to="/recentlyVisited">
                    Title: {article.title.length > 40 ? 'Title to long, show detailed statistics.' : article.title}
                </Link>
                <br/>
                <a href={article.url}>
                    Url: {article.url.length > 40 ? 'Link to long, click here to follow it.' : article.url}
                </a>
                <hr/>
            </div>
        ))}
    </div>
);

export default RecentlyVisitedArticles;