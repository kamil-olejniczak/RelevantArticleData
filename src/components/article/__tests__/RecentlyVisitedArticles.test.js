import * as React from "react";
import RecentlyVisitedArticles from "../RecentlyVisitedArticles";
import renderer from 'react-test-renderer';
import {Link, MemoryRouter} from 'react-router-dom';
import {render, shallow} from 'enzyme';
import {firstRecentlyVisitedSite, secondRecentlyVisitedSite} from '../../../utils/__fixtures__/fakeData';

describe('RecentlyVisitedArticles tests', () => {
    let wrapper;
    const getAnchorContent = () => wrapper.find('a').first().html();
    const componentWithData = data => (
        <RecentlyVisitedArticles recentlyVisited={data} clearData={fakeData.clearData}/>
    );
    const routerOverComponent = data => (
        <MemoryRouter>
            {componentWithData(data)}
        </MemoryRouter>
    );

    beforeEach(() => {
        wrapper = shallow(componentWithData(fakeData.recentlyVisitedImproper));
    });

    it('renders correctly', () => {
        const component = renderer.create(routerOverComponent(fakeData.recentlyVisitedImproper)).toJSON();

        expect(component).toMatchSnapshot();
    });

    it('should render links to articles with their titles', () => {
        expect(wrapper.find('a')).toHaveLength(2);
        expect(wrapper.find(Link)).toHaveLength(2);
    });

    it('should use function to clear local storage when clicking on span', () => {
        const wrapper = shallow(componentWithData(fakeData.recentlyVisitedImproper));

        wrapper.find('#resetStats').simulate('click');

        expect(fakeFunction).toHaveBeenCalled();
    });

    it('should render \'Link to long, click here to follow it.\', when url is to long', () => {
        expect(getAnchorContent()).toContain('Url: Link to long, click here to follow it.');
    });

    it('should render proper url when its length is smaller or equal 40 chars', () => {
        wrapper = shallow(componentWithData(fakeData.recentlyVisitedProper));
        const expectedContent = 'Url: ' + fakeData.recentlyVisitedProper[0].url

        expect(getAnchorContent()).toContain(expectedContent);
    });

    /* there is a bug with shallow, wrapping component in <MemoryRouter> doesn't get rid of
    'Invariant Violation: You should not use <Link> outside a <Router>'
    Using render instead of shallow replaces Link with <a> tag */
    it('should render \'Title to long, show detailed statistics.\', when title is to long', () => {
        wrapper = render(routerOverComponent(fakeData.recentlyVisitedImproper));

        expect(getAnchorContent()).toContain('Title: Title to long, show detailed statistics.');
    });

    it('should render proper title when its length is smaller or equal 40 chars', () => {
        wrapper = render(routerOverComponent(fakeData.recentlyVisitedProper));

        const expectedContent = 'Title: ' + fakeData.recentlyVisitedProper[0].title;
        expect(getAnchorContent()).toContain(expectedContent);
    });
});

const fakeFunction = jest.fn();
const fakeData = {
    recentlyVisitedProper: [firstRecentlyVisitedSite, secondRecentlyVisitedSite],
    recentlyVisitedImproper: [
        {
            ...firstRecentlyVisitedSite,
            title: 'SomeTitle, to long to be displayed in div, additional length',
            url: 'SomeUrl, to long to be displayed in div, additional length',
        },
        {
            ...secondRecentlyVisitedSite,
            title: 'SomeTitle2, to long to be displayed in div, additional length',
            url: 'SomeUrl2, to long to be displayed in div, additional length',
        }
    ],
    clearData: fakeFunction
};
