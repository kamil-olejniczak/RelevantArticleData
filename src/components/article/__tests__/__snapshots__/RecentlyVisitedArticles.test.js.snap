// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentlyVisitedArticles tests renders correctly 1`] = `
<div
  className="RecentlyVisitedArticles"
>
  <span
    className="header"
  >
    Recently Visited Articles:
  </span>
  <span
    id="resetStats"
    onClick={[Function]}
  >
    Reset stats
  </span>
  <div>
    <a
      href="/recentlyVisited"
      onClick={[Function]}
    >
      Title: 
      Title to long, show detailed statistics.
    </a>
    <br />
    <a
      href="SomeUrl, to long to be displayed in div, additional length"
    >
      Url: 
      Link to long, click here to follow it.
    </a>
    <hr />
  </div>
  <div>
    <a
      href="/recentlyVisited"
      onClick={[Function]}
    >
      Title: 
      Title to long, show detailed statistics.
    </a>
    <br />
    <a
      href="SomeUrl2, to long to be displayed in div, additional length"
    >
      Url: 
      Link to long, click here to follow it.
    </a>
    <hr />
  </div>
</div>
`;
